include/thread_pool/thread_pool.hpp

```diff
@@ -75,88 +91,129 @@ class ThreadPoolImpl {
+        it->reset(new Worker<Task, Queue>(options.busyWaitOptions(), options.queueSize()));
 
-    for(size_t i = 0; i < m_workers.size(); ++i)
-    {
-        m_workers[i]->start(i, &m_workers);
-    }
-}
+    // Initialize all worker threads.
+    for (size_t i = 0; i < m_workers.size(); ++i)
+        m_workers[i]->start(i, m_workers, m_idle_workers, m_num_busy_waiters);
 
-template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>::ThreadPoolImpl(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
-{
-    *this = rhs;
+    m_rouser.start(m_workers, m_idle_workers, m_num_busy_waiters);
 }
 
 template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>::~ThreadPoolImpl()
+inline GenericThreadPool<Task, Queue>::GenericThreadPool(GenericThreadPool&& rhs) noexcept
 {
-    for (auto& worker_ptr : m_workers)
-    {
-        worker_ptr->stop();
-    }
+    *this = std::move(rhs);
 }
 
 template <typename Task, template<typename> class Queue>
-inline ThreadPoolImpl<Task, Queue>&
-ThreadPoolImpl<Task, Queue>::operator=(ThreadPoolImpl<Task, Queue>&& rhs) noexcept
+inline GenericThreadPool<Task, Queue>& GenericThreadPool<Task, Queue>::operator=(GenericThreadPool&& rhs) noexcept
 {
     if (this != &rhs)
     {
+        m_idle_workers = std::move(rhs.m_idle_workers);
         m_workers = std::move(rhs.m_workers);
+        m_rouser = std::move(rhs.m_rouser);
         m_next_worker = rhs.m_next_worker.load();
+        m_num_busy_waiters = rhs.m_num_busy_waiters.load();
     }
+
     return *this;
 }
 
+template <typename Task, template<typename> class Queue>
+inline GenericThreadPool<Task, Queue>::~GenericThreadPool()
+{
+    m_rouser.stop();
+
+    for (auto& worker_ptr : m_workers)
+        worker_ptr->stop();
+}
+
 template <typename Task, template<typename> class Queue>
 template <typename Handler>
-inline bool ThreadPoolImpl<Task, Queue>::tryPost(Handler&& handler)
+inline bool GenericThreadPool<Task, Queue>::tryPost(Handler&& handler)
 {
-    return getWorker().tryPost(std::forward<Handler>(handler));
+    // This section of the code increases the probability that our thread pool
+    // is fully utilized (num active workers = argmin(num tasks, num total workers)).
+    // If there aren't busy waiters, let's see if we have any idling threads. 
+    // These incur higher overhead to wake up than the busy waiters.
+    if (m_num_busy_waiters.load(std::memory_order_acquire) == 0)
+    {
+        auto result = m_idle_workers.tryEmptyAny();
+        if (result.first)
+        {
+            auto success = m_workers[result.second]->tryPost(std::forward<Handler>(handler));
+            m_workers[result.second]->wake();
+
+            return success;
+        }
+    }
+
+    // No idle threads. Our threads are either active or busy waiting
+    // Either way, submit the work item in a round robin fashion.
+    if (!getWorker().tryPost(std::forward<Handler>(handler)))
```

Comment:
should we retry the next worker is this fails? Should we retry as many times as there are threads in the pool?  If so, will we need another getWorker() function variant that guarantees retrieving the next thread in the round robin?

Reply:
I'd like to include post retrying in another Pull Request. I have a ticket for it > COM-466. It would (hopefully 😉) be the last feature that we'd need to add to this thread pool for it to be consumable.